### DISCUSSION TITLE
Send data to Webservice at time intervals

### DISCUSSION THREAD

**Discussion by AFumi39** [zepp-health/zeppos-forum#270] [Developer Feedback]
Hello to everyone,
I'm trying to develop something (for personal use, I will publish the code on github, but not the app on the store) that can read sensor data from Amazfit Balance (ZOS 3.0) every X minutes and send all to external endpoint (my WS on a remote server).
I did some test building a Device App that do an httpRequest() to my endpoint, and it works well (using ZML Library). Then I proceeded adding a time-based app-service that reads sensors last value every 10 minutes, and it works well.

Then, I'm currently stuck at this point, because I can't do the httpRequest() (or even the request()) inside the app-service. It give me a "SBEngine: TypeError: not a function" error.
BLE functionality are in the "availability list" of app-service, so I'm expecting that this communication is possible. But I can't figure a way to do it...

Maybe there's some method that permits to "send a message" to "something" that receive the message and do a httpRequest()?
Or maybe there's a way to develop a side-service that get sensor data and sends data to remote endpoint, at regular intervals?

If you have suggestions, please tell me so I can try it.

Thank you

**Comment by silver-zepp** [comment]
Hey. We are currently working on an implementation of this feature without third-party workarounds. There's no ETA for now.

Also, not all the BLE functionality is available, however. Your case seems to rely on the device's timer and that timer cannot work without the UI. But since you are using a Background Service, there's no UI. We will bring some updates into our docs to clarify this part.

The current suggestion would be to use hmBLE directly or a MessageBuilder for this case. You also might have to rely on AOD to make sure that the device's internal timer always works @AFumi39

**Reply by AFumi39** [reply]
@silver-zepp thank you for your reply.
I'm referring to time_service.js example https://github.com/zepp-health/zeppos-samples/blob/OS3.0/application/3.0/3.0-feature/app-service/time_service.js for my test. So you are saying that without AOD it may not works well, right? But so, when the screen reactivates (even without AOD) it should works as expected, right?

I will look into hmBLE and MessageBuilder to see what I can do.

Thank you

**Reply by silver-zepp** [reply]
The time service should work fine despite the UI state. What I was referring to, is when you rely on a side service, or MessageBuilder in particular (which is what ZML is built on) that one might not work for some cases as it relies on the device's internal timer which isn't accessible without the UI.

**Reply by AFumi39** [reply]
I noticed that the documentation of v3 does not match with the example code. I'm now referring to ToDo List example of v2 to see the BT communication.
Thanks

**Reply by silver-zepp** [reply]
Our docs are actually open-sourced now: https://github.com/zepp-health/zeppos-docs
If you see a discrepancy, feel free to contribute! 👋 
Or let us know which part doesn't match so we can fix it.

**Reply by AFumi39** [reply]
> The time service should work fine despite the UI state. What I was referring to, is when you rely on a side service, or MessageBuilder in particular (which is what ZML is built on) that one might not work for some cases as it relies on the device's internal timer which isn't accessible without the UI.

Ah, ok. Thank you, now it's clear 🤦‍♂️
Thank you for the clarification. I just have to try then.


> Our docs are actually open-sourced now: https://github.com/zepp-health/zeppos-docs If you see a discrepancy, feel free to contribute! 👋 Or let us know which part doesn't match so we can fix it.

It differs because the page https://docs.zepp.com/docs/samples/app/toDoList/ (and also the page https://docs.zepp.com/docs/guides/best-practice/bluetooth-communication/ ) refers to /shared folder (and various js) that are not presents in the 3.0 version. Also it refers to Side Service that is not present in 3.0, but only in 2.0.
The 3.0 version rely only on ZML Lib and use no Side Service.

I'll try to fix it when I'll have time, if it will not already fixed.

Thanks

**Reply by Anduin2017** [reply]
Same question here. I'm also trying to call HTTP API in a background (time based) service. The first thing bring up to my mind is to call the "Side Service" on phone. However seems no sample for 3.0 to request side service in an app-service.

**Reply by AFumi39** [reply]
@Anduin2017  See my response below

**Comment by AFumi39** [comment]
> Same question here. I'm also trying to call HTTP API in a background (time based) service. The first thing bring up to my mind is to call the "Side Service" on phone. However seems no sample for 3.0 to request side service in an app-service.

This is the working version for me. Needs a "nightly" version of ZML library (not sure if it's been released as stable yet). It works flawlessy for more than 2 months now.
[hello-world-test3_example.zip](https://github.com/zepp-health/zeppos-forum/files/14343397/hello-world-test3_example.zip)

You just need to replace the webservice url with yours, and change the data structure

**Reply by Anduin2017** [reply]
OH MY GOD! It just works!!!!!!

I found the key point in your zip:

* Upgrade zml to 0.0.23
* Wrap AppService with a BasePage inside (I have no idea if it's expected and no sample mentioned it, but it seems to be the key)

You really should make it an open source project!

**Reply by AFumi39** [reply]
Good, I'm glad to know that it works for you too!! 😊

I faced the same problems as you, tried many different ways and spent many days (maybe weeks) after this. But with the help and support of some developers (especially the ZML developer, that fixed a bug with 0.0.23 version; otherwise, there were no way it would works now) finally managed to complete it ^_^

PS for anyone that need it: the example code is not really "clean", it's dirtied from many attempts to find out the reasons of crash and other things. Just take the structure and use it as "reference" of working app if you have problem

**Reply by Anduin2017** [reply]
I archived the code as a repo and wrote a quick doc about it here: https://gitlab.aiursoft.cn/anduin/sleepagent/
